File: Backend/services/e-sign-service/scripts/export-vsign-uat-excel.js
/**
 * Export completed VSign UAT transactions to Verasys Excel format (Sr No, DATE, TXN ID).
 * Uses Node for MongoDB; invokes Python openpyxl for .xlsx write.
 *
 * Usage:
 *   node scripts/export-vsign-uat-excel.js
 *   node scripts/export-vsign-uat-excel.js --out "C:/Users/DELL/Desktop/50 eSign Transaction Format.xlsx"
 */

require('dotenv').config({ path: require('path').join(__dirname, '..', '..', '..', '.env') });

const fs = require('fs');
const path = require('path');
const { spawnSync } = require('child_process');
const mongoose = require('mongoose');
const SignatureTransaction = require('../models/signatureTransactions');

function parseArgs() {
  const args = process.argv.slice(2);
  let out = 'C:\\Users\\DELL\\Desktop\\50 eSign Transaction Format.xlsx';
  let limit = 50;
  for (let i = 0; i < args.length; i += 1) {
    if (args[i] === '--out' && args[i + 1]) {
      out = args[i + 1];
      i += 1;
    } else if (args[i] === '--limit' && args[i + 1]) {
      limit = Number(args[i + 1]);
      i += 1;
    }
  }
  return { out: path.resolve(out), limit };
}

async function main() {
  const { out, limit } = parseArgs();
  const mongoUri = process.env.MONGO_URI || 'mongodb://127.0.0.1:27017/draftnsign';
  await mongoose.connect(mongoUri);

  const all = await SignatureTransaction.find({}).sort({ txn: 1 }).lean();
  const completed = [];

  for (const t of all) {
    if (t.signedFilePath && fs.existsSync(t.signedFilePath)) {
      const created = t._id.getTimestamp();
      completed.push({
        txn: String(t.txn),
        date: created.toISOString(),
      });
    }
    if (completed.length >= limit) break;
  }

  await mongoose.disconnect();

  const tmpJson = path.join(__dirname, `.vsign-excel-export-${Date.now()}.json`);
  fs.writeFileSync(tmpJson, JSON.stringify({ out, rows: completed, targetRows: 50 }));

  const py = `
import json, sys
from pathlib import Path
from datetime import datetime
from openpyxl import load_workbook, Workbook

data = json.loads(Path(sys.argv[1]).read_text(encoding='utf-8'))
out = Path(data['out'])
rows = data['rows']
target = data.get('targetRows', 50)

if out.exists():
    wb = load_workbook(out)
    ws = wb.active
else:
    wb = Workbook()
    ws = wb.active
    ws.title = 'Sheet1'
    ws.cell(1, 1, 'Sr. No.')
    ws.cell(1, 2, 'DATE')
    ws.cell(1, 3, 'TXN ID')

for i in range(target):
    r = i + 2
    ws.cell(r, 1, i + 1)
    if i < len(rows):
        dt = datetime.fromisoformat(rows[i]['date'].replace('Z', '+00:00')).replace(tzinfo=None)
        ws.cell(r, 2, dt)
        txn = rows[i]['txn']
        ws.cell(r, 3, int(txn) if str(txn).isdigit() else txn)
    else:
        ws.cell(r, 2, None)
        ws.cell(r, 3, None)

wb.save(out)
print(f'Wrote {len(rows)} completed txns to {out}')
`;

  const pyFile = path.join(__dirname, `.vsign-excel-write-${Date.now()}.py`);
  fs.writeFileSync(pyFile, py);
  const result = spawnSync('python', [pyFile, tmpJson], { encoding: 'utf-8' });
  fs.unlinkSync(tmpJson);
  fs.unlinkSync(pyFile);

  if (result.status !== 0) {
    console.error(result.stderr || result.stdout);
    process.exit(1);
  }

  console.log(result.stdout.trim());
  console.log(`Genuine completed: ${completed.length} / 50`);
  completed.forEach((r, i) => console.log(`  ${i + 1}. ${r.txn}`));
  if (completed.length < 50) {
    console.log(`\nRemaining: ${50 - completed.length} — each needs real Aadhaar OTP on esignuat.vsign.in`);
  }
}

main().catch((err) => {
  console.error(err.message);
  process.exit(1);
});

File: Backend/services/e-sign-service/scripts/export-vsign-uat-excel.py
#!/usr/bin/env python3
"""
Export completed VSign UAT transactions into Verasys Excel format:
  Sr. No. | DATE | TXN ID

Only rows with signed PDF on disk are included (genuine completed txns).

Usage:
  python scripts/export-vsign-uat-excel.py
  python scripts/export-vsign-uat-excel.py --out "C:/Users/DELL/Desktop/50 eSign Transaction Format.xlsx"
  python scripts/export-vsign-uat-excel.py --limit 50 --merge "C:/Users/DELL/Desktop/50 eSign Transaction Format.xlsx"
"""

from __future__ import annotations

import argparse
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

try:
    from openpyxl import load_workbook
except ImportError:
    print("Install: pip install openpyxl", file=sys.stderr)
    sys.exit(1)

try:
    from pymongo import MongoClient
    from bson.objectid import ObjectId
except ImportError:
    print("Install: pip install pymongo", file=sys.stderr)
    sys.exit(1)


def load_env_mongo_uri(service_root: Path) -> str:
    env_path = service_root.parent.parent / ".env"
    uri = "mongodb://127.0.0.1:27017/draftnsign"
    if env_path.exists():
        for line in env_path.read_text(encoding="utf-8").splitlines():
            if line.startswith("MONGO_URI="):
                uri = line.split("=", 1)[1].strip()
                break
    return uri


def fetch_completed_txns(mongo_uri: str, limit: int) -> list[dict]:
    client = MongoClient(mongo_uri)
    db = client.get_default_database()
    coll = db["signaturetransactions"]
    rows: list[dict] = []

    for doc in coll.find({}).sort("_id", -1):
        signed_path = doc.get("signedFilePath") or ""
        if not signed_path or not os.path.isfile(signed_path):
            continue
        oid = doc.get("_id")
        created = oid.generation_time if isinstance(oid, ObjectId) else datetime.now(timezone.utc)
        if created.tzinfo is None:
            created = created.replace(tzinfo=timezone.utc)
        local_dt = created.astimezone()
        rows.append(
            {
                "txn": str(doc.get("txn") or ""),
                "date": local_dt.replace(tzinfo=None),
            }
        )
        if len(rows) >= limit:
            break

    client.close()
    rows.sort(key=lambda r: r["txn"])
    return rows


def write_excel(out_path: Path, txns: list[dict], target_rows: int = 50) -> None:
    if out_path.exists():
        wb = load_workbook(out_path)
        ws = wb.active
    else:
        from openpyxl import Workbook

        wb = Workbook()
        ws = wb.active
        ws.title = "Sheet1"
        ws.cell(1, 1, "Sr. No.")
        ws.cell(1, 2, "DATE")
        ws.cell(1, 3, "TXN ID")

    for i in range(target_rows):
        row_idx = i + 2
        if i < len(txns):
            ws.cell(row_idx, 1, i + 1)
            ws.cell(row_idx, 2, txns[i]["date"])
            ws.cell(row_idx, 3, int(txns[i]["txn"]) if txns[i]["txn"].isdigit() else txns[i]["txn"])
        else:
            ws.cell(row_idx, 1, i + 1)
            ws.cell(row_idx, 2, None)
            ws.cell(row_idx, 3, None)

    wb.save(out_path)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--out",
        default=r"C:\Users\DELL\Desktop\50 eSign Transaction Format.xlsx",
        help="Output Excel path",
    )
    parser.add_argument("--limit", type=int, default=50)
    args = parser.parse_args()

    service_root = Path(__file__).resolve().parent.parent
    mongo_uri = load_env_mongo_uri(service_root)
    txns = fetch_completed_txns(mongo_uri, args.limit)

    out_path = Path(args.out)
    write_excel(out_path, txns, target_rows=50)

    print(f"Excel updated: {out_path}")
    print(f"Genuine completed transactions: {len(txns)} / 50")
    for i, t in enumerate(txns, 1):
        print(f"  {i}. {t['date']}  txn={t['txn']}")
    if len(txns) < 50:
        print(
            f"\nNeed {50 - len(txns)} more: complete OTP on esignuat.vsign.in for each new envelope."
        )
        print("Bulk txnref script does NOT count — each txn needs real Aadhaar OTP + signed PDF.")


if __name__ == "__main__":
    main()
